File: testcode/cpp/DebugTest.cpp
#include "DebugTest.hpp"
#include <sstream>
#include <thread>

// DebugBase 実装
void DebugBase::addChild(std::shared_ptr<DebugBase> child) {
    children.push_back(child);
}

// DebugTestA 実装
DebugTestA::DebugTestA() {
    className = "DebugTestA";
    startTime = std::chrono::steady_clock::now();
}

void DebugTestA::execute() {
    std::cout << "Starting " << className << " execution" << std::endl;
    processLevel1();
    for (auto& child : children) {
        child->execute();
    }
}

std::string DebugTestA::getDebugInfo() const {
    auto endTime = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
    std::stringstream ss;
    ss << className << " execution time: " << duration.count() << "ms";
    return ss.str();
}

void DebugTestA::processLevel1() { processLevel2(); }
void DebugTestA::processLevel2() { processLevel3(); }
void DebugTestA::processLevel3() { processLevel4(); }
void DebugTestA::processLevel4() { processLevel5(); }
void DebugTestA::processLevel5() { processLevel6(); }
void DebugTestA::processLevel6() { processLevel7(); }
void DebugTestA::processLevel7() {
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    std::cout << className << " reached level 7" << std::endl;
}

// DebugTestB 実装
DebugTestB::DebugTestB() {
    className = "DebugTestB";
    startTime = std::chrono::steady_clock::now();
}

void DebugTestB::execute() {
    std::cout << "Starting " << className << " execution" << std::endl;
    processLevel1();
    for (auto& child : children) {
        child->execute();
    }
}

std::string DebugTestB::getDebugInfo() const {
    auto endTime = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
    std::stringstream ss;
    ss << className << " execution time: " << duration.count() << "ms";
    return ss.str();
}

void DebugTestB::processLevel1() { processLevel2(); }
void DebugTestB::processLevel2() { processLevel3(); }
void DebugTestB::processLevel3() { processLevel4(); }
void DebugTestB::processLevel4() { processLevel5(); }
void DebugTestB::processLevel5() { processLevel6(); }
void DebugTestB::processLevel6() { processLevel7(); }
void DebugTestB::processLevel7() {
    std::this_thread::sleep_for(std::chrono::milliseconds(150));
    std::cout << className << " reached level 7" << std::endl;
}

// DebugTestC 実装
DebugTestC::DebugTestC() {
    className = "DebugTestC";
    startTime = std::chrono::steady_clock::now();
}

void DebugTestC::execute() {
    std::cout << "Starting " << className << " execution" << std::endl;
    processLevel1();
    for (auto& child : children) {
        child->execute();
    }
}

std::string DebugTestC::getDebugInfo() const {
    auto endTime = std::chrono::steady_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
    std::stringstream ss;
    ss << className << " execution time: " << duration.count() << "ms";
    return ss.str();
}

void DebugTestC::processLevel1() { processLevel2(); }
void DebugTestC::processLevel2() { processLevel3(); }
void DebugTestC::processLevel3() { processLevel4(); }
void DebugTestC::processLevel4() { processLevel5(); }
void DebugTestC::processLevel5() { processLevel6(); }
void DebugTestC::processLevel6() { processLevel7(); }
void DebugTestC::processLevel7() {
    std::this_thread::sleep_for(std::chrono::milliseconds(200));
    std::cout << className << " reached level 7" << std::endl;
}

// DebugTestFacade 実装
DebugTestFacade::DebugTestFacade() {
    testA = std::make_shared<DebugTestA>();
    testB = std::make_shared<DebugTestB>();
    testC = std::make_shared<DebugTestC>();
}

void DebugTestFacade::runAllTests() {
    std::cout << "Running all tests..." << std::endl;
    testA->execute();
    testB->execute();
    testC->execute();
}

void DebugTestFacade::runTestA() {
    std::cout << "Running Test A..." << std::endl;
    testA->execute();
}

void DebugTestFacade::runTestB() {
    std::cout << "Running Test B..." << std::endl;
    testB->execute();
}

void DebugTestFacade::runTestC() {
    std::cout << "Running Test C..." << std::endl;
    testC->execute();
} 

File: testcode/python/debug_test.py
import time
from abc import ABC, abstractmethod
from typing import List
from dataclasses import dataclass, field
from datetime import datetime

class DebugBase(ABC):
    def __init__(self):
        self.class_name = self.__class__.__name__
        self.start_time = time.time()
        self.children: List['DebugBase'] = []

    def add_child(self, child: 'DebugBase') -> None:
        self.children.append(child)

    @abstractmethod
    def execute(self) -> None:
        pass

    def get_debug_info(self) -> str:
        duration = (time.time() - self.start_time) * 1000  # Convert to milliseconds
        return f"{self.class_name} execution time: {duration:.2f}ms"

class DebugTestA(DebugBase):
    def execute(self) -> None:
        print(f"Starting {self.class_name} execution")
        self._process_level1()
        for child in self.children:
            child.execute()

    def _process_level1(self) -> None: self._process_level2()
    def _process_level2(self) -> None: self._process_level3()
    def _process_level3(self) -> None: self._process_level4()
    def _process_level4(self) -> None: self._process_level5()
    def _process_level5(self) -> None: self._process_level6()
    def _process_level6(self) -> None: self._process_level7()
    def _process_level7(self) -> None:
        time.sleep(0.1)
        print(f"{self.class_name} reached level 7")

class DebugTestB(DebugBase):
    def execute(self) -> None:
        print(f"Starting {self.class_name} execution")
        self._process_level1()
        for child in self.children:
            child.execute()

    def _process_level1(self) -> None: self._process_level2()
    def _process_level2(self) -> None: self._process_level3()
    def _process_level3(self) -> None: self._process_level4()
    def _process_level4(self) -> None: self._process_level5()
    def _process_level5(self) -> None: self._process_level6()
    def _process_level6(self) -> None: self._process_level7()
    def _process_level7(self) -> None:
        time.sleep(0.15)
        print(f"{self.class_name} reached level 7")

class DebugTestC(DebugBase):
    def execute(self) -> None:
        print(f"Starting {self.class_name} execution")
        self._process_level1()
        for child in self.children:
            child.execute()

    def _process_level1(self) -> None: self._process_level2()
    def _process_level2(self) -> None: self._process_level3()
    def _process_level3(self) -> None: self._process_level4()
    def _process_level4(self) -> None: self._process_level5()
    def _process_level5(self) -> None: self._process_level6()
    def _process_level6(self) -> None: self._process_level7()
    def _process_level7(self) -> None:
        time.sleep(0.2)
        print(f"{self.class_name} reached level 7")

class DebugTestFacade:
    def __init__(self):
        self.test_a = DebugTestA()
        self.test_b = DebugTestB()
        self.test_c = DebugTestC()

    def run_all_tests(self) -> None:
        print("Running all tests...")
        self.test_a.execute()
        self.test_b.execute()
        self.test_c.execute()

    def run_test_a(self) -> None:
        print("Running Test A...")
        self.test_a.execute()

    def run_test_b(self) -> None:
        print("Running Test B...")
        self.test_b.execute()

    def run_test_c(self) -> None:
        print("Running Test C...")
        self.test_c.execute()

if __name__ == "__main__":
    facade = DebugTestFacade()
    facade.run_all_tests() 